server/src/main/java/org/elasticsearch/plugins/scanners/NamedComponentReader.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.plugins.scanners;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.core.Strings;
+import org.elasticsearch.plugins.PluginBundle;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
+
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import static java.util.Collections.emptyMap;
+import static org.elasticsearch.xcontent.XContentType.JSON;
+
+/**
+ * Reads named components declared by a plugin in a cache file.
+ * Cache file is expected to be present in plugin's lib directory
+ *
+ * The content of a cache file is a JSON representation of a map where:
+ * keys -> name of the extensible interface (a class/interface marked with @Extensible)
+ * values -> a map of name to implementation class name
+ */
+public class NamedComponentReader {
+
+    private Logger logger = LogManager.getLogger(NamedComponentReader.class);
+    private static final String NAMED_COMPONENTS_FILE_NAME = "named_components.json";
+    /**
+     * a registry of known classes marked or indirectly marked (extending marked class) with @Extensible
+     */
+    private final ExtensiblesRegistry extensiblesRegistry;
+
+    public NamedComponentReader() {
+        this.extensiblesRegistry = ExtensiblesRegistry.INSTANCE;
+    }
```

Comment:
let's chain this constructor, so we only have one substantive constructor body, e.g. `this(ExtensiblesRegistry.INSTANCE)`